core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableProject.java

```diff
@@ -86,4 +100,74 @@ public Result implement(EnumerableRelImplementor implementor, Prefer pref) {
     // EnumerableCalcRel is always better
     throw new UnsupportedOperationException();
   }
+
+  @Override public Pair<RelTraitSet, List<RelTraitSet>> passThroughTraits(
+      RelTraitSet required) {
+    RelCollation collation = required.getCollation();
+    if (collation == null) {
+      return null;
+    }
+
+    final Mappings.TargetMapping map =
+        RelOptUtil.permutationIgnoreCast(
+            getProjects(), getInput().getRowType());
+
+    for (RelFieldCollation rc : collation.getFieldCollations()) {
+      if (!isCollationOnTrivialExpr(map, rc)) {
+        return null;
+      }
+    }
+
+    final RelCollation newCollation = collation.apply(map);
+    return Pair.of(required, ImmutableList.of(required.replace(newCollation)));
+  }
+
+  @Override public Pair<RelTraitSet, List<RelTraitSet>> deriveTraits(
+      final RelTraitSet childTraits, final int childId) {
+    RelCollation collation = childTraits.getCollation();
+    if (collation == null) {
+      return null;
+    }
+
+    final Mappings.TargetMapping map =
+        RelOptUtil.permutationIgnoreCast(
+            getProjects(), getInput().getRowType());
```

Comment:
You can't use the same mapping. This is used to map parent to child. What you want is map child position to parent position.

Reply:
Let's not consider `cast` in this PR:
```java
int maxField = Math.max(inputFieldCount, projects.size());
Mappings.TargetMapping mapping = Mappings
    .create(MappingType.FUNCTION, maxField, maxField);
for (int i = 0; i < projects.size(); i++) {
  RexNode exp = projects.get(i);
  if (exp instanceof RexInputRef) {
    mapping.set(((RexInputRef) exp).getIndex(), i);
  }
}
```

If you check the body of `permutationIgnoreCast`, they are just reversed mapping.

Reply:
Thanks. Updated. 

Later we could add util functions for such reverse operations.

Reply:
There is. You can use `Project.getPartialMapping()` instead.

Reply:
If you don't want to support cast in `derive`, then the check `isCollationOnTrivialExpr` is not need anymore.

Reply:
Let me dig into this part of code and see what is going on.

Reply:
Added the missing CAST part. 

It was due to how tests are written. E.g. the cast leads to non-monotonic behaves the same of a normal RexCall. Thus even CAST misses, nothing was broken.

I used debugger to confirm now that CAST test cases hit the correct code path.